integration/bip0009_test.go

```diff
@@ -37,6 +50,255 @@ func assertVersionBit(r *rpctest.Harness, t *testing.T, hash *chainhash.Hash, bi
+// status is the equivalent of the passed state.
+func assertSoftForkStatus(r *rpctest.Harness, t *testing.T, forkKey string, state blockchain.ThresholdState) {
+	// Convert the expected threshold state into the equivalent
+	// getblockchaininfo RPC status string.
+	status, err := thresholdStateToStatus(state)
+	if err != nil {
+		_, _, line, _ := runtime.Caller(1)
+		t.Fatalf("assertion failed at line %d: unable to convert "+
+			"threshold state %v to string", line, state)
+	}
+
+	info, err := r.Node.GetBlockChainInfo()
+	if err != nil {
+		t.Fatalf("failed to retrieve chain info: %v", err)
+	}
+
+	// Ensure the key is available.
+	desc, ok := info.Bip9SoftForks[forkKey]
+	if !ok {
+		_, _, line, _ := runtime.Caller(1)
+		t.Fatalf("assertion failed at line %d: softfork status for %q "+
+			"is not in getblockchaininfo results", line, forkKey)
+	}
+
+	// Ensure the status it the expected value.
+	if desc.Status != status {
+		_, _, line, _ := runtime.Caller(1)
+		t.Fatalf("assertion failed at line %d: softfork status for %q "+
+			"is %v instead of expected %v", line, forkKey,
+			desc.Status, status)
+	}
+}
+
+// testBIP0009 ensures the BIP0009 soft fork mechanism follows the state
+// transition rules set forth by the BIP for the provided soft fork key.  It
+// uses the regression test network to signal support and advance through the
+// various threshold states including failure to achieve locked in status.
+//
+// See TestBIP0009 for an overview of what is tested.
+//
+// NOTE: This only differs from the exported version in that it accepts the
+// specific soft fork deployment to test.
+func testBIP0009(t *testing.T, forkKey string, deploymentID uint32) {
+	// Initialize the primary mining node with only the genesis block.
+	r, err := rpctest.New(&chaincfg.RegressionNetParams, nil, nil)
+	if err != nil {
+		t.Fatalf("unable to create primary harness: %v", err)
+	}
+	if err := r.SetUp(false, 0); err != nil {
+		t.Fatalf("unable to setup test chain: %v", err)
+	}
+	defer r.TearDown()
+
+	// *** ThresholdDefined ***
+	//
+	// Assert the chain height is the expected value and the soft fork
+	// status starts out as defined.
+	assertChainHeight(r, t, 0)
+	assertSoftForkStatus(r, t, forkKey, blockchain.ThresholdDefined)
+
+	// *** ThresholdDefined part 2 - 1 block prior to ThresholdStarted ***
+	//
+	// Generate enough blocks to reach the height just before the first
+	// state transition.
+	//
+	// NOTE: This is two blocks before the confirmation window because the
+	// getblockchaininfo RPC reports the status for the block AFTER the
+	// current one.  All of the heights below are thus offset by one to
+	// compensate.
+	//
+	// Assert the chain height is the expected value and soft fork status is
+	// still defined and did NOT move to started.
+	confirmationWindow := r.ActiveNet.MinerConfirmationWindow
+	for i := uint32(0); i < confirmationWindow-2; i++ {
+		_, err := r.GenerateAndSubmitBlock(nil, vbLegacyBlockVersion,
```

Comment:
I modified the comment a bit to specifically call out why it's using a version that does not signal support.